darling/src/libsystem_kernel/emulation/src/xnu_syscall/bsd/impl/misc/clonefile.c

```diff
@@ -1,13 +1,224 @@
+
+                dest_fd = next_fd;
+                continue;
+            }
+
+            if (path == NULL) {
+                filename = token;
+                break;
+            }
+
+            if (errno != ENOENT) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            };
+
+            if (sys_mkdirat(dest_fd, token, 0755) < 0) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            }
+
+            next_fd = sys_openat(dest_fd, token, BSD_O_RDONLY | BSD_O_DIRECTORY | BSD_O_CLOEXEC, st.st_mode & 0777);
+
+            if (next_fd < 0) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            }
+
+            if (dest_fd != LINUX_AT_FDCWD) sys_close(dest_fd);
+
+            dest_fd = next_fd;
+        }
+    }
+
+    int real_dest_fd = sys_openat(dest_fd, filename, open_flags, 0644);
+
+    if (errno != ENOENT) {
+        return -errno;
+    }
+
+    size_t remaining;
+    off_t off_in = 0;
+    off_t off_out = 0;
+    size_t to_copy;
+    size_t copied;
+
+    if (st.st_size > 0) {
+        remaining = (size_t)st.st_size;
+        off_in = 0;
+        off_out = 0;
+
+        while (remaining > 0) {
+            to_copy = remaining;
+
+            if (to_copy > MAX_COPY_CHUNK) to_copy = MAX_COPY_CHUNK;
+
+#warning use sys_copy_file_range instead of syscall
+            copied = LINUX_SYSCALL(__NR_copy_file_range, src_fd, &off_in, real_dest_fd, &off_out, to_copy, 0);
+
+            if (copied < 0) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            }
+
+            if (copied == 0) {
+                break;
+            }
+
+            if (copied > remaining) {
+                remaining = 0;
+            } else {
+                remaining -= copied;
+            }
+        }
+    }
+
+    if (!(flags & CLONE_NOOWNERCOPY)) {
+        if (sys_fchown(real_dest_fd, st.st_uid, st.st_gid) < 0) {
+            saved_errno = errno;
+            ret = -saved_errno;
+            goto out_close;
+        }
+    }
+
+    if (sys_fchmod(real_dest_fd, st.st_mode & 07777) < 0) {
+        saved_errno = errno;
+        ret = -saved_errno;
+        goto out_close;
+    }
+
+    struct timespec times[2];
+    times[0] = st.st_atimespec;
+    times[1] = st.st_mtimespec;
+#ifdef HAVE_FUTIMENS
+    if (futimens(real_dest_fd, times) < 0) {
+        saved_errno = errno;
+        ret = -saved_errno;
+        goto out_close;
+    }
+#else
+    struct timeval tv[2];
+    tv[0] = (struct timeval){
+        .tv_sec = times[0].tv_sec,
+        .tv_usec = times[0].tv_nsec / 1000,
+    };
+    tv[1] = (struct timeval){
+        .tv_sec = times[1].tv_sec,
+        .tv_usec = times[1].tv_nsec / 1000,
+    };
+
+    if (futimes(real_dest_fd, tv) < 0) {
```

Comment:
Stick with the direct syscall implementation instead. Don't forget to get replace the errno.

```suggestion
    if (sys_futimes(real_dest_fd, tv) < 0) {
```